Reqnroll/Formatters/RuntimeSupport/FeatureLevelCucumberMessages.cs

```diff
@@ -1,21 +1,113 @@
 using Io.Cucumber.Messages.Types;
+using Reqnroll.Formatters.PayloadProcessing;
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
 
 namespace Reqnroll.Formatters.RuntimeSupport;
 
 /// <summary>
 /// This class is used at Code Generation time to provide serialized representations of the Source, GherkinDocument, and Pickles 
 /// to be used at runtime.
 /// </summary>
-public class FeatureLevelCucumberMessages(Func<Source> source, Func<GherkinDocument> gherkinDocument, Func<IEnumerable<Pickle>> pickles)
+public class FeatureLevelCucumberMessages : IFeatureLevelCucumberMessages
 {
-    private Lazy<Source> _source = new Lazy<Source>(source);
-    private Lazy<GherkinDocument> _gherkinDocument = new Lazy<GherkinDocument>(gherkinDocument);
-    private Lazy<IEnumerable<Pickle>> _pickles = new Lazy<IEnumerable<Pickle>>(pickles);
-
-    public bool HasMessages => Source is not null && GherkinDocument is not null && Pickles is not null;
-    public Source Source { get => _source.Value; }
-    public GherkinDocument GherkinDocument { get => _gherkinDocument.Value; }
-    public IEnumerable<Pickle> Pickles { get => _pickles.Value; }
+    private Lazy<IEnumerable<Envelope>> _embeddedEnvelopes;
+    internal readonly int _expectedEnvelopeCount;
+    private Lazy<Source> _source;
+    private Lazy<GherkinDocument> _gherkinDocument;
+    private Lazy<IEnumerable<Pickle>> _pickles;
+
+    public FeatureLevelCucumberMessages(string resourceNameOfEmbeddedMessages, int envelopeCount)
+    {
+        if (string.IsNullOrEmpty(resourceNameOfEmbeddedMessages))
+            throw new ArgumentNullException(nameof(resourceNameOfEmbeddedMessages));
+
+        var assm = Assembly.GetCallingAssembly();
+        _embeddedEnvelopes = new Lazy<IEnumerable<Envelope>>(() => ReadEnvelopesFromAssembly(assm, resourceNameOfEmbeddedMessages));
```

Comment:
I would change the type of `_embeddedEnvelopes` to `Lazy<Envelope[]>` or `Lazy<IReadOnlyCollection<Envelope>>`. Otherwise the readers (like me) falsely think that it re-reads the steam every time when it is accessed. (For that reason you normally rarely use `IEnumerable` and `Lazy` together, because the first is for pushing the caching responsibility to the caller, while the other keeps the caching on the producer side, so they are somewhat contradictory.)